### PR TITLE
Check access permissions to WebPortals host

### PR DESCRIPTION
**Pull request by danigargar**
#### Type Of Change 
- [ ] Small bug fix
- [x] New feature or enhancement
- [ ] Breaking change

#### Checklist:
- [x] Commits are named and have tag following [commit rules](https://github.com/irontec/ivozprovider/blob/bleeding/doc/dev/en/commits.md)
- [x] Commits are split per component (schema, portal/platform, kamusers, agis, ..)
- [x] Changes have been tested locally
- [ ] Fixes an existing issue (Fixes #XXXX) 
- [ ] Upport from existing Pull request #XXXX

#### Description
Implemented login checks to restrict access using the WebPortal URL assigned to a company or brand, ensuring it is relevant to the user or admin.

### CODE REVIEW

**Review by ironArt3mis — CHANGES_REQUESTED** [review]
Functional review required

File: library/Ivoz/Provider/Domain/Service/Administrator/AdminLoginChecker.php [hunk 1]
```diff
@@ -2,13 +2,20 @@
 
 namespace Ivoz\Provider\Domain\Service\Administrator;
 
+use Symfony\Component\HttpFoundation\RequestStack;
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Service/Administrator/AdminLoginChecker.php:5]: Request is an infraestruture concept, it should not be part of a domain service

File: library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        if ($webPortal->getUrlType() === WebPortalInterface::URLTYPE_GOD) {
+            return;
+        }
+
+        if (is_null($webPortal->getCompany())) {
+            $this->assertBrandsMatch($user, $webPortal) ;
+        } else {
+            $this->assertCompaniesMatch($user, $webPortal);
+        }
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php:37]:
```
        if (is_null($webPortal->getCompany())) {
            $this->assertBrandsMatch($user, $webPortal) ;
            return;
        }

        $this->assertCompaniesMatch($user, $webPortal);
```
Would be better

File: library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    private function assertBrandsMatch(AdministratorInterface $user, WebPortalInterface $webPortal): void
+    {
+        $userBrandId = $user->getBrand()?->getId();
+        $webPortalBrandId = $webPortal->getBrand()?->getId();
+
+        $hasBrands = !(is_null($userBrandId) || is_null($webPortalBrandId));
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php:45]:
```suggestion
        $brandMustMatch = !(is_null($userBrandId) || is_null($webPortalBrandId));
```

File: library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        $this->throwAccessDeniedException();
+    }
+
+    private function assertCompaniesMatch(AdministratorInterface $user, WebPortalInterface $webPortal): void
+    {
+        $userCompanyId = $user->getCompany()?->getId();
+        $webPortalCompanyId = $webPortal->getCompany()?->getId();
+
+        $hasCompanies = !(is_null($userCompanyId) || is_null($webPortalCompanyId));
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php:58]:
```suggestion
        $companyMustMatch = !(is_null($userCompanyId) || is_null($webPortalCompanyId));
```

File: library/Ivoz/Provider/Domain/Service/User/UserLoginChecker.php [hunk 5]
```diff
@@ -3,32 +3,45 @@
...
+use Symfony\Component\HttpFoundation\RequestStack;
 use Symfony\Component\Security\Core\Exception\CustomUserMessageAccountStatusException;
 use Symfony\Component\Security\Core\User\UserCheckerInterface;
 use Symfony\Component\Security\Core\User\UserInterface as SymfonyUserInterface;
 use Ivoz\Provider\Domain\Model\User\UserInterface;
 
 class UserLoginChecker implements UserCheckerInterface
 {
+    public function __construct(
+        private RequestStack $requestStack,
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Service/User/UserLoginChecker.php:16]: This is infra concept

**Review by mmadariaga — CHANGES_REQUESTED** [review]

File: web/rest/platform/features/provider/company/getCompany.feature [hunk 6]
```diff
@@ -73,7 +73,19 @@ Feature: Manage companies
...
+                "type": "wholesale",
+                "name": "Wholesale Company 2",
+                "maxDailyUsage": 1000000,
+                "currentDayUsage": -1,
+                "billingMethod": "postpaid",
+                "id": 6,
+                "invoicing": {
+                    "nif": "12345699-Z"
+                },
+                "domainName": null
```
Comment (mmadariaga) [inline web/rest/platform/features/provider/company/getCompany.feature:87]: This could be a little bit more different from company#5

**Review by mmadariaga — CHANGES_REQUESTED** [review]

File: library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+final class AssertWebPortalAccessible
+{
+    public function __construct(
+        private WebPortalRepository $webPortalRepository
+    ) {
+    }
+
+    public function execute(AdministratorInterface $user, string $domain): void
+    {
+        if ($user->isSuperAdmin()) {
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php:19]: Does this apply the rule `No se aplican limitaciones si la parte host de la URL está dado de alta como URL de un nivel distinto al que pertenece el administrador que se está logando  (e.g. dominio de brand logándose con /client con credenciales de cliente: no hay limitaciones)`?

**Review by mmadariaga — CHANGES_REQUESTED** [review]
Some suggestions

File: library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php (hunk 7 above)
Comment (danigargar) [reply library/Ivoz/Provider/Domain/Service/Administrator/AssertWebPortalAccessible.php:19]: That wording is a bit confusing. The rule is "no one should see a WebPortal theme owned by a company (brand) they don't belong to". 

So if you are a SuperAdmin no limitation applies to you.

File: library/Ivoz/Provider/Domain/Service/Administrator/AdminLoginChecker.php (hunk 1 above)
Comment (danigargar) [reply library/Ivoz/Provider/Domain/Service/Administrator/AdminLoginChecker.php]: I have moved it to a service within the infrastructure, injected via a domain interface. We didn't have anything like this before, so I'm uncertain about how to name the components and organize the folders.

I'm really looking forward to hearing your opinion about that.

**Review by mmadariaga — APPROVED** [review]

File: library/Ivoz/Provider/Domain/HostnameGetter.php [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Ivoz\Provider\Domain;
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Domain/HostnameGetter.php:3]:
```suggestion
namespace Ivoz\Provider\Domain\Service;
```
Would be better

File: library/Ivoz/Provider/Infrastructure/Api/Hostname/GetHostname.php [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace Ivoz\Provider\Infrastructure\Api\Hostname;
```
Comment (mmadariaga) [inline library/Ivoz/Provider/Infrastructure/Api/Hostname/GetHostname.php:3]: I would put this out of Api namespace. I don't see any relationship between "get host name" and API

**Review by mmadariaga — APPROVED** [review]

**Review by ironArt3mis — APPROVED** [review]

**Pull request merged**